### ISSUE TITLE
FSE: Allow the ability to Save Draft and Schedule changes

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#29575]
## What problem does this address?

Currently, there isn't a way to save drafts or schedule changes to the site editor experience. Right now, this is an option with the customizer and can be super useful when overhauling a part of the site but wanting to wait for a quieter traffic time to launch changes. Here's what it looks like in the customizer:

![image (6)](https://user-images.githubusercontent.com/26996883/110047327-eaf4f880-7d0a-11eb-84e1-2cf59f0350cf.png)

This came up as part of the [FSE Outreach Program's second call for testing](https://make.wordpress.org/test/2021/02/18/fse-program-testing-call-2-build-a-homepage-with-site-editing-blocks/#comment-1367). Of note, this touches a bit on what's raised in this issue as well with saving template drafts but it felt important to separate out this concept into its own issue: https://github.com/WordPress/gutenberg/issues/27905 

## What is your proposed solution?

Allow for an option to schedule changes during the saving flow in the site editor.

**Comment by jameskoster** [comment]
I believe scheduling would need to be per entity, and the UI would have to clarify that somehow. This begs a question: should it be possible to have different schedules for different entities?

**Comment by courane01** [comment]
@coreymckrill is working on a revisions plugin/tool for Learn and likely Docs that may be useful here. https://github.com/coreymckrill/revisions-extended. It’s available for testing in Make Testing.

**Comment by coreymckrill** [comment]
My understanding is that FSE templates are actually posts of the `wp_template` post type, so using revisions to save drafts, schedule for later, etc. could probably work the same way the above plugin does.

**Comment by javierarce** [comment]
I've been exploring some interfaces to schedule changes and save drafts. Following @jameskoster idea, here's a version that would allow saving changes per entity. 

From the sidebar, users can mark several entities to be published on a future date (or as a draft), and receive a link to preview all those changes together. The publication settings for each entity could also be managed independently. 

https://user-images.githubusercontent.com/4933/113328036-00fed480-931c-11eb-96ed-c0988f792582.mov

This prototype borrows some interface ideas from [this issue](https://github.com/WordPress/gutenberg/issues/29577)

[Figma prototype](https://www.figma.com/proto/Lr1b05mTHl9TFdNBgKUMob/FSE%3A-Playground?node-id=506%3A10354&scaling=min-zoom&page-id=506%3A8769)

**Comment by jameskoster** [comment]
Nice work on a 🤯 problem @javierarce !

The "Save draft" flow gave me a moment of confusion. It may just be a labelling issue, but saving something _that is already published_ as a draft sounds... contradictory somehow? Could there be two separate drafting actions based on whether the entity is already published?

1. If the entity is already published, there is an option to un-publish / switch to a draft status
2. If the entity is not yet published, there is an option to not publish it on save

"Discard changes" and "Reset to initial state", are these different? In this context they feel like they're describing the same action. If the reset action here is a revert (IE solving [this](https://github.com/WordPress/gutenberg/issues/29687) issue) then I would question whether this is the right place to include such an action.

On a similar note, seeing the un-published changes seems like a nice addition, but I wonder about the placement. I wonder if you tried placing them in the "Template part" tab of the [isolated view](https://github.com/WordPress/gutenberg/issues/29148)? That way we could present a consistent UI for this feature in template editing, isolated template _part_ editing, and perhaps even isolated reusable block editing as well.

**Comment by jasmussen** [comment]
Great prototype, great followup comments by Jay!

I'd add that it does seem like there's potential in attaching the scheduling state as a property of each sub-item on the list that gets highlighted for saving/updating. I'd probably not use the ellipsis on hover, though — seems like the goal there is to find something more discoverable but still visually reduced. 

And yes, finding a superb label for the save button seems like a separate challenge on its own, to Jay's point. Can we have multiple labels depending on the state of the document and what gets picked up for saving?

**Comment by javierarce** [comment]
My answers for @jameskoster [comments](https://github.com/WordPress/gutenberg/issues/29575#issuecomment-812079896):

>  It may just be a labelling issue, but saving something _that is already published_ as a draft sounds... contradictory somehow

In my head the action is saying: "save the last changes I've made as a draft", but I understand what you mean. Maybe a more appropriate name could be "Mark as draft"? Or the main "Save" button could say "Publish" instead? 
 
> Could there be two separate drafting actions based on whether the entity is already published?  
> 1. If the entity is already published, there is an option to un-publish / switch to a draft status
> 2. If the entity is not yet published, there is an option to not publish it on save

I think that the second action is already achievable using the checkbox, right? And, in my opinion, the first one is more complex because in order to be consistent, the option would have to be available for entities that haven't gone through the drafting phase too… and that would make it a bit confusing. 
 
> I wonder if you tried placing them in the "Template part" tab of the isolated view? That way we could present a consistent UI for this feature in template editing, isolated template part editing, and perhaps even isolated reusable block editing as well.

I didn't think about that, but I guess it would make sense to have the information about the draft/schedule status in both places.

> "Discard changes" and "Reset to initial state", are these different? In this context they feel like they're describing the same action.

For me, they behave differently. After clicking in any of those options, the template in question will be removed from the list and its changes will be reverted to the previous state (Discard changes) or to the default state (Reset initial state). 

I understand that the placement could be not ideal (because of the destructive nature of those options) but I think it could work. At the end of this video for an older and different prototype, I explain how this could work (the design of the popover is a bit different from the one in this ticket):

https://user-images.githubusercontent.com/4933/113690471-bab6c600-96cb-11eb-856f-43aea6d76cfc.mov

**Comment by javierarce** [comment]
Answering @jasmussen comments [here](https://github.com/WordPress/gutenberg/issues/29575#issuecomment-813213433):

> I'd add that it does seem like there's potential in attaching the scheduling state as a property of each sub-item on the list that gets highlighted for saving/updating. 

That's a good idea, I think that could work too and it would reduce the redundancy of the column after scheduling a change or creating a draft.

![compact](https://user-images.githubusercontent.com/4933/113687015-2e56d400-96c8-11eb-992d-78e5b99b7ce6.png)


> I'd probably not use the ellipsis on hover, though — seems like the goal there is to find something more discoverable but still visually reduced.

I understand your concern, but on the other hand, I think that drafting and scheduling are more advanced actions that could be accessed through the ellipsis. In order to make them a little bit more discoverable, maybe we could hint at that possibility through the description on top?

**Comment by jasmussen** [comment]
> I understand your concern, but on the other hand, I think that drafting and scheduling are more advanced actions that could be accessed through the ellipsis. In order to make them a little bit more discoverable, maybe we could hint at that possibility through the description on top?

Absolutely! For something that will likely be used so very rarely, it's crucial that it doesn't weigh down the interface and add mental overhead. The challenge here is that as black text each item looks entirely inert. It seems like if we are going to add some interactivity there, there should be some indication other than the description. 

Could it be tiny versions of the icons we have, shown as extension of the text? I don't love these, but they sort of convey my point:

<img width="311" alt="Screenshot 2021-04-06 at 12 03 31" src="https://user-images.githubusercontent.com/1204802/113694605-44689280-96d0-11eb-8036-6f8163b22c60.png">

**Comment by draganescu** [comment]
This is so cool! 

I'd love to have one big "schedule all changes" place/button/flow instead of having to set the same date for three to n items.

**Comment by critterverse** [comment]
I think this issue is really close to being able to be picked up by a developer based on the flow @javierarce shared in https://github.com/WordPress/gutenberg/issues/29575#issuecomment-812044660.

>I'd probably not use the ellipsis on hover, though — seems like the goal there is to find something more discoverable but still visually reduced.

I like @jasmussen's suggestion above and there were some nice alternatives for this shared in https://github.com/WordPress/gutenberg/issues/29575#issuecomment-813997566.

Another option could be to show the ellipses in it's current placement but in a lighter gray until the row is hovered over:

https://user-images.githubusercontent.com/77359364/154168551-9e88c11f-0b0e-49bd-9b5d-cc0345394cc5.mov

One side note: these options all seem to work a lot better without the expanding/collapsing arrows that the individual panels have currently — they add a lot more visual noise when paired with this new persistent element and I'm not sure how useful that functionality is in this context.

**Comment by westonruter** [comment]
I believe it would make more sense (and more common) to allow the user to schedule _all_ the changes together, rather than to have a way to schedule each separate entity separately. The Customizer's approach is to have scheduling for all of the changes being made, not separate schedules for each change. Being able to schedule each change separately feels like a power user feature that would not be used by most.

**Comment by critterverse** [comment]
>I believe it would make more sense (and more common) to allow the user to schedule all the changes together, rather than to have a way to schedule each separate entity separately

Good to know, seems like there have been requests for both ways. Curious if anyone sees value in the ability to schedule by panel grouping — Site, Template, etc? Just thinking out loud because it would resolve some of the trickier aspects of this UI if we could use a single ellipses "more" menu at the top of each panel.

**Comment by westonruter** [comment]
For reference, here's a video showing the Customizer UI for scheduling all changes made in the current session:

https://user-images.githubusercontent.com/134745/154374654-2df3f598-70d7-45b3-8ac2-fed1a12c6974.mov

More screenshots can be seen in the [Changeset Scheduling](https://make.wordpress.org/core/2017/11/03/new-features-and-enhancements-with-customizer-changesets-in-4-9/#:~:text=programmatically%20trash%20changesets.-,Changeset%20Scheduling,-Since%20the%20values) section of [New Features and Enhancements with Customizer Changesets in 4.9](https://make.wordpress.org/core/2017/11/03/new-features-and-enhancements-with-customizer-changesets-in-4-9/).

When working on the Customizer changeset scheduling feature, I don't recall there being any discussion of the need to schedule individual setting changes separately. That being said, there was actually a way to do so but still at the session level and not the setting level. There is the concept of a “[Changeset Branching Mode](https://make.wordpress.org/core/2017/11/03/new-features-and-enhancements-with-customizer-changesets-in-4-9/#:~:text=restored%20(see%C2%A0%2339031).-,Changeset%20Branching%20Mode,-In%20spite%20of)” which allows you to have multiple pending changesets in parallel (which required a PHP filter to opt-in). So you could have one changeset that modifies the site title and schedule that for publishing and then another changeset that modifies the site tagline and schedule that separately. A reason why we didn't default to have the ability multiple changesets pending at a time is (1) there wasn't a clear user need, (2) it would complicate the UI by needing to introduce a way to manage changesets, and (3) there would be the possibility for two separate pending changesets to have conflicts in the changes that they are intending to make (which would necessitate some UI to warn users of such conflicts as well).

**Comment by javierarce** [comment]
I want to share a prototype of a global scheduler system that I created last year (it was my first attempt to solve this problem, but I ended up working on a more granular version). It's obviously a bit outdated and will also require some polishing, but maybe it can work as an alternative.

https://user-images.githubusercontent.com/4933/154666146-caffc364-83f2-47ef-81cf-0c44ba763338.mp4


[Link to the prototype](https://www.figma.com/proto/Lr1b05mTHl9TFdNBgKUMob/FSE%3A-Playground?page-id=386%3A0&node-id=430%3A12034&viewport=352%2C48%2C0.44&scaling=min-zoom&starting-point-node-id=430%3A4557)

**Comment by jasmussen** [comment]
Nice, thanks for sharing Javi! 

I wonder if a split-button treatment with a dropdown chevron might work instead of the cog 🤔  — there's something about the cog being used for the inspector that trips me up a little. Something like this perhaps?

<img width="163" alt="Frame 865" src="https://user-images.githubusercontent.com/1204802/154666841-235af65f-ac7c-4cef-a50e-479300915365.png">

**Comment by javierarce** [comment]
Yes, that would totally work and looks much better than the cog (I guess I used it to replicate the way the Customizer does it)

**Comment by coreymckrill** [comment]
Here's a screencast of [the plugin](https://github.com/WordPress/revisions-extended/) we built for WordPress.org that allows for scheduling updates to published posts, which incidentally uses the same dropdown style button:

https://d.pr/v/f2623L

**Comment by jasmussen** [comment]
Nice, thanks for sharing Corey. Definitely looks similar.

**Comment by courtneyr-dev** [comment]
I'd love to see (and may have earlier suggested) making use of this plugin when pursuing feature parity with scheduling in Customizer. It's been stable and useful for Training & Docs.

**Comment by javierarce** [comment]
I've updated my previous proposal with a split button, as @jasmussen suggested. 

https://user-images.githubusercontent.com/4933/168110102-eda17124-9fbe-4b4c-b548-e8e723262804.mov


Something else I've added from the previous iteration is a warning notice in the case the user schedules a post for publication but then decides to save a draft instead. 

![image](https://user-images.githubusercontent.com/4933/168110730-cc49412d-f264-429b-8d9c-3946b5f8d0a8.png)

You can find the prototype [here](https://www.figma.com/proto/RdY36KONIC8XmQsRYn1Na7/Multi-entity-Saving%3A-Allow-the-ability-to-Save-Draft-and-Schedule-changes?page-id=1%3A2&node-id=1%3A2&viewport=-3017%2C-7835%2C1.81&scaling=min-zoom&starting-point-node-id=502%3A3034). And here's [the design](https://www.figma.com/file/RdY36KONIC8XmQsRYn1Na7/Multi-entity-Saving-Allow-the-ability-to-Save-Draft-and-Schedule-changes).

**Comment by jameskoster** [comment]
I find saving drafts in the site editor to be a little confusing. What happens if I edit post content then save a draft, does that update the post status to draft (and unpublish it) or just store those changes somewhere? If the changes are stored, what happens when I go to edit that post, can I pull the changes in?

**Comment by mtias** [comment]
This is looking good! In the context of templates and the site editor we should avoid saying things like "this post is scheduled" because users can get confused. Expressing it as "these template updates" or "these design changes" would be better.

**Comment by javierarce** [comment]
> In the context of templates and the site editor we should avoid saying things like "this post is scheduled" because users can get confused.

Oh, yes, you're totally right, I don't know what I was thinking when I wrote that 🤦‍♂️  I'll fix it.

**Comment by javierarce** [comment]
Based on the feedback I got from @shaunandrews I've made several changes to the design:

* The button now says “Update” instead of “Save”.
* The split bottom tooltip shows all the available actions with a tick next to the selected one.
* Uses the [upcoming Date popover](https://github.com/WordPress/gutenberg/pull/41097) instead of dropdowns.
* The schedule and draft information sections appear on top of the list of template areas
* The checkboxes appear on the right side of the sidebar, to be consistent with some changes we are implementing

Here's a video of the [new prototype](https://www.figma.com/proto/RdY36KONIC8XmQsRYn1Na7/Multi-entity-Saving%3A-Allow-the-ability-to-Save-Draft-and-Schedule-changes?page-id=554%3A7187&node-id=554%3A8232&viewport=1001%2C438%2C0.35&scaling=min-zoom&starting-point-node-id=554%3A8232):

https://user-images.githubusercontent.com/4933/169809702-b41e7ad1-1440-437d-8a87-659a66521509.mov

And some notes about how this design presents the different saving options:

![image](https://user-images.githubusercontent.com/4933/169809781-75c84390-9c5f-4ba1-a4c2-e61370191e06.png)

**Comment by draganescu** [comment]
This is so cool ❤️ 

I wonder:

- will the schedule updates feature only be available for editing templates, not when editing individual template parts?
- just checking, we assume only one set of scheduled changes for template is available at a time, right?
- I can't grasp from the mockup what happens if edit a scheduled template and it's parts, will I have to "reschedule" and just select the same date/time?

Using the post revision system there is a situation I can't figure out:

1. Edit the header in home template -> schedule
2. Edit the header in single template -> schedule

This means we need to keep two versions of the scheduled changes for the header template part, right? Or will scheduling changes for template parts in a template affect another template (unlikely)?

**Comment by javierarce** [comment]
Ah, those are great questions I should have included in my design.  

> will the schedule updates feature only be available for editing templates, not when editing individual template parts?

I think only at the template level, otherwise it could get very confusing and complicated to manage.

> I can't grasp from the mockup what happens if edit a scheduled template and it's parts, will I have to "reschedule" and just select the same date/time?

If you edit a scheduled template the default option will be to add the changes to the current scheduling, but you'll be able to: a) adjust the scheduling b) save as a draft c) save immediately.

> This means we need to keep two versions of the scheduled changes for the header template part, right? Or will scheduling changes for template parts in a template affect another template (unlikely)?

Here's a mental excercise of what I think could happen (this is quite complex, so maybe I'm totally wrong and/or maybe there's a simpler way 😅):

- Template A contains Template Part 1 and 2.
- Template B contains Template Part 1, 2, and 3.

1. You first go to Template A and make several changes (on the template itself and also on the template parts). You schedule the changes to be published tomorrow.

2. You then go to edit Template B. You'll see Template B and the updated (but not just published) versions of Template Parts 1 and 2 [a question here is if we could indicate that they have scheduled changes]

3. You make some changes to Template B and also to Template Parts 1, 2, and 3. Then schedule the changes to be released in two weeks.

Here's what I think will happen:

- After 1 day, Template A is updated with the changes made for that template but also with the modifications for Template Part 1, and 2 that were introduced when updating Template B.

- After 2 weeks, Template B gets updated with its changes. Since Template Part 1 and 2 were already updated, users will only notice the change in Template B and Template Part 3.

**Comment by javierarce** [comment]
And then there's also [@jameskoster question](
https://github.com/WordPress/gutenberg/issues/29575#issuecomment-1125152490):

> What happens if I edit post content then save a draft, does that update the post status to draft (and unpublish it) or just store those changes somewhere? If the changes are stored, what happens when I go to edit that post, can I pull the changes in?

Should we exclude the content changes from that flow? And… is that something we can technically do?

**Comment by pablohoneyhoney** [comment]
Perhaps the new button with a drop down could use some more padding and finesse. @javierarce 
How could we improve that whole top header with the button and the close icon, like position when the it slides in. 
Do you think the large close icon works well in scale and also in context of a smaller version in the settings inspector?

**Comment by pablohoneyhoney** [comment]
This is a separate issue or consideration, but this effort made me think how disconnected the action button (top right) and the snackbar validation (bottom left) are.

**Comment by talldan** [comment]
Are there concerns over how easy it'll be to break a site by switching a template part to draft? I did try to read through the whole thread and couldn't see this addressed, but I may have missed it.

If I'm editing template A and switch my header template part to draft, then that may accidentally unpublish the same template part in templates B, C, D etc. It seems like that would be incredibly easy to do, and have quite a bad result.

Scheduling is less risky, but similarly a user may be unaware that scheduling a change in a template part will affect multiple templates.

I also think it's worth thinking through the use cases for scheduling something. I think seasonal or promotional content is a big reason, but there are many others.

In this situation, rather than editing a template part in place, I wonder if a way to fork or create variations of a template part might be more interesting. A user could perhaps be able to schedule a seasonal variation to go live, but also be able to schedule an end date when the original version of the template part is reverted back to.

**Comment by jameskoster** [comment]
I don't even know that scheduling _is_ less risky. For example it would be easy to schedule changes to a template in the site editor, then go and save changes to that template in the post editor... what happens when the scheduled changes go live... what if there are conflicts? Drafting / scheduling that involves content or templates still feels like a can of worms to me.

**Comment by mtias** [comment]
Scheduling is essentially a fork through revisions. Generally, editors load the scheduled version so you are always editing the latest. It gets complex when the thing that is scheduled is an update because you also want to be able to see what is published. They are workable problems, though.
